### PR TITLE
Optimize ReplaceAliasByActualDefinitionPass

### PR DESCRIPTION
**Pull request by ajb-in**
| Q | A |
| --- | --- |
| Branch? | 2.3 |
| Bug fix? | yes (performance) |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Previous implementation passed over every definition for every alias (n*m runtime).
New implementation passes once over all aliases and once over all definitions (n+m).
Also removing needless "restart" logic.

### CODE REVIEW

**Comment by ajb-in** [comment]
Note: patch applies to master with minor modification (tested).

**Comment by lstrojny** [comment]
Improves container dump times from ~40s to ~16s in real world applications.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 1]
```diff
@@ -39,98 +38,74 @@ public function process(ContainerBuilder $container)
...
+            $newId = $replacements[$refId];
+            $arguments[$k] = new Reference($newId, $argument->getInvalidBehavior());
+            $this->compiler->addLogMessage($this->formatter->formatUpdateReference($this, $defId, $refId, $newId));
         }
 
         return $arguments;
     }
 
-    private function updateFactoryServiceReference($factoryService, $currentId, $newId)
+    private function updateFactoryReferenceId($replacements, $referenceId)
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: What about adding a doc block here?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 2]
```diff
@@ -39,98 +38,74 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: IMO ``$treatedAliases`` or ``$managedAliases`` would be better.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 3]
```diff
@@ -39,98 +38,74 @@ public function process(ContainerBuilder $container)
...
+            if (array_key_exists($aliasId, $seenAliases)) {
+                continue;
+            }
+            $seenAliases[$aliasId] = true;
             try {
                 $definition = $container->getDefinition($aliasId);
             } catch (InvalidArgumentException $e) {
-                throw new InvalidArgumentException(sprintf('Unable to replace alias "%s" with actual definition "%s".', $id, $alias), null, $e);
+                $msg = sprintf('Unable to replace alias "%s" with actual definition "%s".', $defId, $alias);
+                throw new InvalidArgumentException($msg, null, $e);
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: should be inlined.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 4]
```diff
@@ -39,98 +38,74 @@ public function process(ContainerBuilder $container)
...
-                    $this->compiler->addLogMessage($this->formatter->formatUpdateReference($this, $this->sourceId, $currentId, $newId));
-                }
+                $arguments[$k] = $this->updateArgumentReferences($replacements, $defId, $argument);
+                continue;
+            }
+            if (!($argument instanceof Reference)) {
+                continue;
             }
+            $refId = (string) $argument;
+            if (!array_key_exists($refId, $replacements)) {
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: ``isset`` is faster

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 5]
```diff
@@ -39,98 +38,74 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $defId => $alias) {
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: IMO ``$id`` was fine.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 1 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: On it :+1:

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 2 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: This keeps tracks of all aliases seen in this pass, whether they need replacement (i.e. point to a private def) or not.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 3 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: :ok_hand:

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 4 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Can I safely assume it'll never be null?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 6]
```diff
@@ -39,98 +38,74 @@ public function process(ContainerBuilder $container)
...
             $aliasId = (string) $alias;
-
+            if (array_key_exists($aliasId, $replacements)) {
+                $container->setAlias($defId, $replacements[$aliasId]);
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: The goal is to remove the initial service when possible so please remove this line...

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 5 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: I found it ambiguous since that section handles several ids simultaneously.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 6 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Actually I think that's equivalent to what the original code was doing: all subsequent aliases that pointed to the old (removed) alias are changed to point to the new one.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 4 above)
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: This looks strange to me that a service id could be ``null`` but as ``array_key_exists`` is used in ``ContainerBuilder`` let this as is.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 1 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Done

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 3 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Done

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 4 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Makes sense. Changed it to `isset`

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 7]
```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
...
-                $arguments[$k] = $this->updateArgumentReferences($argument, $currentId, $newId);
-            } elseif ($argument instanceof Reference) {
-                if ($currentId === (string) $argument) {
-                    $arguments[$k] = new Reference($newId, $argument->getInvalidBehavior());
-                    $this->compiler->addLogMessage($this->formatter->formatUpdateReference($this, $this->sourceId, $currentId, $newId));
-                }
+                $arguments[$k] = $this->updateArgumentReferences($replacements, $defId, $argument);
+                continue;
+            }
+            if (!($argument instanceof Reference)) {
```
Comment (stloyd) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Should be:
```php
if (!$argument instanceof Reference)

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 6 above)
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Only private services are replaced so they don't need to be available after this optimization.

Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: So, should references to these aliases also be replaced in definitions?

Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Ah sorry you're right, I thought it was doing something else...

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 7 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: That looks as though it said `(!$argument) instanceof Reference)`. I know precedence rules work the other way around, but it still looks visually confusing IMO. I'll change it if it's a big fugly though.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 6 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: So I should leave it that way then?

Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Yeah that's fine :+1:

**Comment by Ener-Getick** [comment]
This looks great :+1:
Thank you @ajb-in :smile:

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 7 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Done :(

**Comment by ajb-in** [comment]
Thanks @Ener-Getick ^_^

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 7 above)
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: @ajb-in this is always written this way in symfony ([see this](https://github.com/symfony/symfony/search?utf8=%E2%9C%93&q=instanceof))

Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Fair enough :beer:

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 2 above)
Comment (javiereguiluz) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: `$processedAliases` ? In any case, I agree with @Ener-Getick: `$seenAliases` seems wrong because of the `seen` verb.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 8]
```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
...
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $defId => $alias) {
             $aliasId = (string) $alias;
-
+            if (isset($replacements[$aliasId])) {
+                $container->setAlias($defId, $replacements[$aliasId]);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: This must respect the visibility of the existing alias (allowing it to be removed in case of a private alias not being referenced anywhere anymore after other optimizations)

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 5 above)
Comment (javiereguiluz) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Short variable names usually reduce legibility: what about using `$definitionId`?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 9]
```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
...
-                    $this->compiler->addLogMessage($this->formatter->formatUpdateReference($this, $this->sourceId, $currentId, $newId));
-                }
+                $arguments[$k] = $this->updateArgumentReferences($replacements, $defId, $argument);
+                continue;
+            }
+            if (!$argument instanceof Reference) {
+                continue;
             }
+            $refId = (string) $argument;
+            if (!isset($replacements[$refId])) {
```
Comment (javiereguiluz) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Same here: what about using `$referenceId`?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 10]
```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
...
+                continue;
+            }
+            $seenAliases[$aliasId] = true;
             try {
                 $definition = $container->getDefinition($aliasId);
             } catch (InvalidArgumentException $e) {
-                throw new InvalidArgumentException(sprintf('Unable to replace alias "%s" with actual definition "%s".', $id, $alias), null, $e);
+                throw new InvalidArgumentException(sprintf('Unable to replace alias "%s" with actual definition "%s".', $defId, $alias), null, $e);
             }
-
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: please don't remove empty lines. there are here for readability of the source code

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 11]
```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
+        $seenAliases = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $defId => $alias) {
             $aliasId = (string) $alias;
-
+            if (isset($replacements[$aliasId])) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: What if you have chained aliased, and the target alias is processed only later ? This would not replace it properly anymore (which makes me think that this case is not properly covered by tests btw)

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 12]
```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
...
+     * Returns the updated reference for the factory service.
+     *
+     * @param array  $replacements Table of aliases to replace
+     * @param string $referenceId  Factory service reference identifier
+     *
+     * @return string
+     */
+    private function updateFactoryReferenceId($replacements, $referenceId)
     {
-        if (null === $factoryService) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: This case must be kept, to avoid issues in case there is a service with an empty string id (which is valid)

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 13]
```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
...
-                $arguments[$k] = $this->updateArgumentReferences($argument, $currentId, $newId);
-            } elseif ($argument instanceof Reference) {
-                if ($currentId === (string) $argument) {
-                    $arguments[$k] = new Reference($newId, $argument->getInvalidBehavior());
-                    $this->compiler->addLogMessage($this->formatter->formatUpdateReference($this, $this->sourceId, $currentId, $newId));
-                }
+                $arguments[$k] = $this->updateArgumentReferences($replacements, $defId, $argument);
+                continue;
+            }
+            if (!$argument instanceof Reference) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: please add empty lines between the different ``if`` blocks for readability

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 2 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: This does not keep track of aliases you saw in the past, but of alias *targets* you saw in the past (which might not be aliases at all)

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 4 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: A service id is always a string

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 11 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: the easiest way to handle it would probably be to restart the handling of aliases to build the table of replacements, keeping the update of Definitions for the end to do it only once.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 2 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: :+1: renaming to `seenAliasTargets`

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 5 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: :+1:

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 8 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: I tried to stick to the original behaviour, namely: remove the definition of the target, point the alias to the definition, and point every subsequent alias with the same target to the first alias.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 9 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: :+1:

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 11 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Restarting the pass would mean quadratic runtime, which would beat the purpose of this patch.
I could do it with a union-find structure though. Could I get a test case to check against please? :3

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 12 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Wouldn't that case be handled normally by the logic?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 10 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Can we compromise on section-breaking comments? (see new commit) :smile:

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 13 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: better?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 11 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: @stof: so the logic should be: if I have a chain `a1 -> a2 -> a3 -> a4 -> d` (with `d` private), that should change to `a1 -> a3`, `a2 -> a3`, `a3 -> d` (with `d` public) and `a4` deleted?

Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: @stof: also, should alias chains that end in a public definition be affected?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 12 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: no, because tried to access ``$replacements[$referenceId]`` will access ``$replacements['']``, as ``null`` is not a valid array key and so get casted string

Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: Oh fun :unamused:
Should I be explicit and `return null;` here or leave it as `return`?

Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: leave it as it is today, due to the Symfony coding standards (even though I personally don't like this particular rule)

Comment (SenseException) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: @stof Why was an empty string defined as valid?

Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: @SenseException there is nothing to do to define a string as valid. there is work to do to forbid it. And given that we have not written code to forbid it, it means that the empty id is accepted by the code today, and so must continue to work (in case someone had the weird idea to use it).

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 11 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: @stof: Looking at the original code again, I think the restart step doesn't do anything anyway. Aliases are replaced only if the definition of their target is marked as private; the process is to mark the definition as public and replace all references to the alias target with references to the alias id (and then removing the definition of the alias target). This process does not introduce any new private definitions, therefore no alias target that was previously seen as having a public definition should have a private one after it. Am I missing something?

**Comment by ajb-in** [comment]
@stof re chains: I played with the existing testcase for 2.3 to see what happens with chained aliases. I ran this code:

        $container = new ContainerBuilder();

        $aDefinition = $container->register('a', '\stdClass');
        $aDefinition->setFactory(array(new Reference('b'), 'createA'));

        $bDefinition = new Definition('\stdClass');
        $bDefinition->setPublic(false);
        $container->setDefinition('b', $bDefinition);

        $container->setAlias('a_alias', 'a');
        $container->setAlias('b_alias', 'b');

        // I added these three lines:
        $container->setAlias('c_alias', 'b');
        $container->setAlias('d_alias', 'c_alias');
        $container->setAlias('e_alias', 'd_alias');

        $this->process($container);

and it 'sploded *on 2.3*:

<pre>There was 1 error:

1) Symfony\Component\DependencyInjection\Tests\Compiler\ReplaceAliasByActualDefinitionPassTest::testProcess
Symfony\Component\DependencyInjection\Exception\InvalidArgumentException: Unable to replace alias "d_alias" with actual definition "c_alias".

/code/symfony/src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php:48
/code/symfony/src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php:63
/code/symfony/src/Symfony/Component/DependencyInjection/Tests/Compiler/ReplaceAliasByActualDefinitionPassTest.php:69
/code/symfony/src/Symfony/Component/DependencyInjection/Tests/Compiler/ReplaceAliasByActualDefinitionPassTest.php:40

Caused by
Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: You have requested a non-existent service "c_alias".</pre>

**Comment by ajb-in** [comment]
It explodes also if I change the extra part to this:

        // I added these two lines:
        $container->setAlias('c_alias', 'b_alias');
        $container->setAlias('d_alias', 'c_alias');

**Comment by tadcka** [comment]
:+1:

**Comment by egils** [comment]
Lovely! :+1:

**Comment by xabbuh** [comment]
@ajb-in Aliases that refer to other aliases should have been resolved in the `ResolveReferencesToAliasesPass` which is run before the `ReplaceAliasByActualDefinitionPass`.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 14]
```diff
@@ -39,98 +38,91 @@ public function process(ContainerBuilder $container)
...
-    private function updateFactoryServiceReference($factoryService, $currentId, $newId)
+    /**
+     * Returns the updated reference for the factory service.
+     *
+     * @param array  $replacements Table of aliases to replace
+     * @param string $referenceId  Factory service reference identifier
+     *
+     * @return string
+     */
+    private function updateFactoryReferenceId($replacements, $referenceId)
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: missing typehint array for `$replacements`

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 15]
```diff
@@ -39,98 +38,91 @@ public function process(ContainerBuilder $container)
...
 
         return $arguments;
     }
 
-    private function updateFactoryServiceReference($factoryService, $currentId, $newId)
+    /**
+     * Returns the updated reference for the factory service.
+     *
+     * @param array  $replacements Table of aliases to replace
+     * @param string $referenceId  Factory service reference identifier
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: string|null as $referenceId can be null. phpdoc of `Definition::getFactoryService` should be fixed similarly.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 16]
```diff
@@ -39,98 +38,91 @@ public function process(ContainerBuilder $container)
...
+     *
+     * @return string
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: string|null

**Comment by Tobion** [comment]
:+1: apart from those small doc fixes.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 17]
```diff
@@ -39,98 +38,91 @@ public function process(ContainerBuilder $container)
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
-        foreach ($container->getAliases() as $id => $alias) {
-            $aliasId = (string) $alias;
-
+        $seenAliasTargets = array();
+        $replacements = array();
+        foreach ($container->getAliases() as $definitionID => $target) {
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: should be ``$definitionId`` as ``$targetId`` is used later

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 18]
```diff
@@ -39,98 +38,91 @@ public function process(ContainerBuilder $container)
...
+                continue;
+            }
+            // Process new target
+            $seenAliasTargets[$targetId] = true;
             try {
-                $definition = $container->getDefinition($aliasId);
+                $definition = $container->getDefinition($targetId);
             } catch (InvalidArgumentException $e) {
-                throw new InvalidArgumentException(sprintf('Unable to replace alias "%s" with actual definition "%s".', $id, $alias), null, $e);
+                throw new InvalidArgumentException(sprintf('Unable to replace alias "%s" with actual definition "%s".', $definitionID, $target), null, $e);
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: why not using ``$targetId`` here?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 17 above)
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: ?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 18 above)
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: agree

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 17 above)
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: the variable is upper-cased, i think it should be changed for consistency.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 19]
```diff
@@ -39,98 +38,91 @@ public function process(ContainerBuilder $container)
...
-     * @param string $currentId The alias identifier
-     * @param string $newId     The identifier the alias points to
+     * @param array  $replacements Table of aliases to replace
+     * @param string $definitionID Identifier of this definition
+     * @param array  $arguments    Where to replace the aliases
      *
      * @return array
      */
-    private function updateArgumentReferences(array $arguments, $currentId, $newId)
+    private function updateArgumentReferences(array $replacements, $definitionID, array $arguments)
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: same `Id` not `ID`

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 17 above)
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: i see, true

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 18 above)
Comment (ajb-in) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: No idea :3 That's what the old code was doing though (using `$alias` instead of `$aliasId`). I'll change it.

**Comment by ajb-in** [comment]
@Tobion @Ener-Getick Did doc and variable name fixes; looks better?

**Comment by ajb-in** [comment]
@xabbuh @stof Then this pass doesn't need to be concerned with chains whatsoever, right?

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 15 above)
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: please fix the phpdoc of the two methods on `Definition` as well.

**Comment by fabpot** [comment]
As this is not a bug fix, it should be merged in master, not 2.3

**Comment by ajb-in** [comment]
@fabpot When I asked in #symfony I was told that it could count as a performance bug, so that's why I went with 2.3; @javiereguiluz seemed to agree since he labeled this PR as 'bug'.
I'll rebase it into master if needed, of course.

**Comment by fabpot** [comment]
Sorry about the confusion. It's true that we've been more willing to merge PRs that do not strictly fix bugs in maintained branches (and I probably merged some myself), but I think this is a mistake. First, we want old branches to be as stable as possible, and merging non-bug fixes can introduce subtle regressions (happened in the past). Then, improving performance in new versions is a great incentive for people to upgrade, which is always a good idea.

/cc @symfony/deciders

**Comment by Tobion** [comment]
I'd prefer to merge refactorings like this one in maitainance branches as it will make maintaining branches easier when they are more similar and do not diverge that much just because of refactorings.

**Comment by Tobion** [comment]
And introducing suble behavior changes should not be an argument. It just means we do not have enough tests (or it's really an edge case that wasn't meant to be like this). E.g. the refactoring in the templating parser actually revealed that people use absolute paths that wasn't intended that way.

**Comment by fabpot** [comment]
@ajb-in Can you rebase on current 2.3? Thanks.

**Comment by weaverryan** [comment]
In general, I'd prefer things like this to be merged into master. If this causes a subtle behavior change that isn't caught by our tests, that should be our problem, not a problem that is discovered by users in a patch update. We need users to have very high confidence in upgrading patch versions (otherwise, they may stay on insecure versions!)

But it's not black and white - a performance issue could be a bug, so we can do anything here.

**Comment by ajb-in** [comment]
Rebased on latest 2.3; also squashed, added a comment, and added a plug to this review on the commit message :3

**Comment by fabpot** [comment]
@ajb-in Can you remove the reference in the commit message? We already have everything in place to automatically get that in Git notes.

**Comment by ajb-in** [comment]
@fabpot Done; just wanted to give some due credit :3

**Comment by fabpot** [comment]
Tests are broken apparently.

**Comment by ajb-in** [comment]
@fabpot I forgot to rename a variable =P
Fixed it now

**Comment by fabpot** [comment]
Thank you @ajb-in.

**Comment by ajb-in** [comment]
:beers: Thanks to everyone!
